packages/vite/src/node/build.ts

```diff
@@ -542,55 +544,75 @@ async function doBuild(
     parallelBuilds.push(bundle)
 
     const generate = (output: OutputOptions = {}) => {
-      return bundle[options.write ? 'write' : 'generate'](
-        buildOutputOptions(output)
-      )
+      return bundle[options.write ? 'write' : 'generate'](output)
     }
 
     if (options.write) {
-      prepareOutDir(outDir, options.emptyOutDir, config)
+      prepareOutDir(outDirs, options.emptyOutDir, config)
     }
 
-    if (Array.isArray(outputs)) {
-      const res = []
-      for (const output of outputs) {
-        res.push(await generate(output))
-      }
-      return res
-    } else {
-      return await generate(outputs)
+    const res = []
+    for (const output of normalizedOutputs) {
+      res.push(await generate(output))
     }
+    return Array.isArray(outputs) ? res : res[0]
   } catch (e) {
     outputBuildError(e)
     throw e
   }
 }
 
+function parseShouldEmptyDirs(outDirs: string[]) {
+  const sortedOutDirs = Array.from(new Set(outDirs))
+    .filter(Boolean)
+    .sort((p, n) => p.length - n.length)
+  const removedIndex = new Set<number>()
+
+  // if there have dirs like ['a/b/c', 'b/c', 'c'], only the `a/b/c` should be emptied
+  for (let i = 0; i < sortedOutDirs.length - 1; ++i) {
+    if (removedIndex.has(i)) continue
+
+    for (let j = i + 1; j < sortedOutDirs.length; ++j) {
+      if (removedIndex.has(j)) continue
+
+      if (sortedOutDirs[i].includes(sortedOutDirs[j])) {
+        removedIndex.add(j)
+      }
+    }
+  }
+
+  return sortedOutDirs.filter((_, i) => !removedIndex.has(i))
+}
+
 function prepareOutDir(
-  outDir: string,
+  outDirs: string[],
   emptyOutDir: boolean | null,
   config: ResolvedConfig
 ) {
-  if (fs.existsSync(outDir)) {
-    if (
-      emptyOutDir == null &&
-      !normalizePath(outDir).startsWith(config.root + '/')
-    ) {
-      // warn if outDir is outside of root
-      config.logger.warn(
-        colors.yellow(
-          `\n${colors.bold(`(!)`)} outDir ${colors.white(
-            colors.dim(outDir)
-          )} is not inside project root and will not be emptied.\n` +
-            `Use --emptyOutDir to override.\n`
+  const shouldEmptyDirs = parseShouldEmptyDirs(outDirs)
+
+  for (const outDir of outDirs) {
+    if (fs.existsSync(outDir)) {
+      if (
+        emptyOutDir == null &&
+        !normalizePath(outDir).startsWith(config.root + '/')
+      ) {
+        // warn if outDir is outside of root
+        config.logger.warn(
+          colors.yellow(
+            `\n${colors.bold(`(!)`)} outDir ${colors.white(
+              colors.dim(outDir)
+            )} is not inside project root and will not be emptied.\n` +
+              `Use --emptyOutDir to override.\n`
+          )
         )
-      )
-    } else if (emptyOutDir !== false) {
-      emptyDir(outDir, ['.git'])
+      } else if (emptyOutDir !== false && shouldEmptyDirs.includes(outDir)) {
+        emptyDir(outDir, ['.git'])
```

Comment:
Thanks for the PR. The refactoring looks good to me. But The `parseShouldEmptyDirs` logic doesn't seem right.
I think we can use the skip list in `emptyDir`. Something like:
```suggestion
      } else if (emptyOutDir !== false) {
        const skipDirs = outDirs.map((dir) => {
          const relative = path.relative(outDir, dir)
          if (
	          relative &&
	          !relative.startsWith('..') &&
	          // We don't need to skip nested dirs
	          path.dirname(relative) === '.' &&
	          !path.isAbsolute(relative)
	        ) {
	          return relative
	        }
        })
        .filter(Boolean)

        emptyDir(outDir, ['.git', ...skipDirs])
```

Reply:
Thanks your suggestion. The original logic will remove the `.git` if the dir is nested. But your suggestion also will remove the `.git` when the dir is nested more then two level. I take more try.

Reply:
But this was always the case, no? `.git` was only respected if it is the root of an `outDir`. I think we should tackle ignoring git at every level in a different PR (and I don't know if it is something we need to support).

Reply:
Maybe I did not describe well (my English not good). The case I said just like there have two dirs: `es` and `es/one/two`, and the `es/one/two/.git` will be removed.

Reply:
But I think this case is rare, so I am not sure should I resolve it in this PR. Hope your suggestion.

Reply:
Oh, I see, you are right. mmm... the logic then is more complex, maybe we should extend `emptyDir` skip list to support nested paths? And then we can avoid the `path.dirname(relative) === '.'` line?

Reply:
My thought is similar, but it seems will make `emptyDir` backs to emptying dirs by recursion. If this is right, I will follow this direction.

Reply:
We could check first the `skip` list and find nested paths, for them, we can skip there `path.dirname` with the regular function and then call `emptyDir` on this dirname. So it is recursive only if you pass a nested skip path, so perf shouldn't be affected

Reply:
ok, I try it.